### ISSUE TITLE
Ramps

### ISSUE THREAD

**Issue by ElliotB256** [TeamAtomECS/AtomECS#26]
Hi @MauriceZeuner  and @d-garrick ,

I added a WIP implementation of ramps using a generic system here: https://github.com/TeamAtomECS/AtomECS/blob/collisions/src/ramp.rs

I intend to define a macro, `rampable`, so you can auto derive the required trait impl.

**Comment by ElliotB256** [comment]
I've added the tests for this and the module works nicely. I will pick the commits off collisions and move them into master as I think it's a useful feature to have.

To add a ramp, implement the `Lerp` trait for your `Component`, add a `Ramp<Component>` to your entity, and add a `RampUpdateSystem<Component>` to your dispatcher.

I am considering changing it from a Lerp trait to a blanket implementation for components with `Add` and `Multiply`.

**Comment by ElliotB256** [comment]
Decided to keep it as Lerp - it doesn't make sense to define eg `Add` or `Multiply` for components such as `CoolingLight`, where there are some fields which shouldn't be combined as these operations would imply.

91aeaa - added a derive macro for `Lerp` which means you don't have to explicitly implement the `Lerp` trait for your types - just use `[#derive(Clone,Lerp)]` for your structs. Note that it's fairly crude - it will just go through the fields and add/multiply by an `f64`, so it can only really work for `f64` or `Vector3<f64>` fields. It also assumes `Clone` is implemented for the type. Still better than nothing.

**Comment by ElliotB256** [comment]
eg, from the test, define a new simple component with a float field:
```    #[derive(Clone, Lerp)]
    struct ALerpComp {
        value: f64,
    }```

Now create a ramp for it:
```        let mut frames = Vec::new();
        frames.push((0.0, ALerpComp { value: 0.0 }));
        frames.push((1.0, ALerpComp { value: 1.0 }));
        frames.push((2.0, ALerpComp { value: 0.0 }));
        let mut ramp = Ramp {
            prev: 0,
            keyframes: frames,
        };```

last, add a `RampUpdateSystem::<ALerpComp>::default()` to your dispatcher

**Comment by ElliotB256** [comment]
ecb564d - added derive to some commonly used components.

**Comment by ElliotB256** [comment]
@MauriceZeuner - I'll merge this to master probably tomorrow if I get time.
@DavidGarrick - I've updated the Rb173 shell example to use this approach for the RF knife

**Comment by MauriceZeuner** [comment]
Are you still planning to merge it to master?

**Comment by ElliotB256** [comment]
yes but I didn't get time yet, hopefully tomorrow

**Comment by MauriceZeuner** [comment]
can we merge the ramps into master?  I'd love to play around with them for dipole trapping and optical transport. Also, I have to understand them - don't quite get the "traits" in Rust, especially that Lerp.

**Comment by ElliotB256** [comment]
Merged in 3bc5246
